File: Python/Proxy/Proxy.py
#! /usr/bin/python
# -*- coding: utf-8 -*-
'''
Proxy Pattern
Author: Kei Nakata
Data: Oct.6.2014
'''
import abc

class Subject(object):
    __metaclass__ = abc.ABCMeta
    @abc.abstractmethod
    def show(self):
        pass

class ConcreteSubject(Subject):
    def show(self):
        print "Hello World"

class Proxy(Subject):
    def __init__(self):
        self.concrete_subject = None

    def show(self):
        if(self.concrete_subject == None):
            self.concrete_subject = ConcreteSubject()
        self.concrete_subject.show()


if __name__ == '__main__':
    a = Proxy()
    a.show()

File: C++/Proxy/Proxy.cpp
/*
 * Proxy Pattern
 * Author: Kei Nakata
 * Date: Oct.6.2014
 * Provide a surrogate or placeholder for another object to control
 * access to it.
 */
#include <iostream>
#include <cstdlib>
#include <memory>
using std::cout;
using std::endl;
using std::unique_ptr;

class Subject
{
public:
  Subject() = default;
  virtual ~Subject() = default;
  virtual void Request() = 0;
};

class ConcreteSubject : public Subject
{
protected:
  friend class Proxy;
  ConcreteSubject() = default;
public:
  virtual ~ConcreteSubject() = default;
  virtual void Request()
  {
    cout << "Hello World!!" << endl;
  }
};

class Proxy : Subject
{
private:
  unique_ptr<ConcreteSubject> a;
public:
  Proxy() = default;
  virtual ~Proxy() = default;

  void Request()
  {
    if(!a) a.reset(new ConcreteSubject);
    a->Request();
  }
};

int main()
{
  Proxy a;
  a.Request();

  return EXIT_SUCCESS;
}
